WordPress/gutenberg
# Site Editor: Browsing sidebar templates menu restructure

## Description

- All Primary and Secondary templates according to the WP Template Hierarchy appear at the root level
- Variable templates are sorted into the "Posts" and "Pages" sections accordingly. IE post-$id and category-$id go in the "Posts" sections, and page-$id goes in the "Pages" section.
- $custom templates go in the "General" section.
- Unused templates (`home`, `index`, `singular`) goes to unused menu
- `page-home` is an exception, it's set to shown up in General

### Unused templates

Q: Why only `home`, `index` and `singular` can be in the Unused templates menu?
We wanted to keep it simple and only primary and secondary templates are moved to Unused menu. Hiding unused templates accurately would be [extremely difficult](https://github.com/WordPress/gutenberg/issues/26264#issuecomment-712180860).

Only 3 templates met these conditions:
- can be superseded by only primary and/or secondary templates
- we don't take dynamic and custom templates into account

`home` met the conditions because it is considered unused if:
- `front-page` template exists
- OR custom page template is shown on front

`index` met the conditions because it is considered unused if:
- `archive`, `404`, `search` template exist
- AND (`single` and `page` exist OR `singular` exists)
- AND (`home` exists OR superseded by `front-page` OR custom page shown on front)

`singular` met the conditions because it is considered unused if:
- `single` and `page` template exist

## How has this been tested?
1. Open site editor
2. Open browsing sidebar
3. Navigate to Templates
4. Make sure Primary and Secondary templates show up here
5. Open Posts - make sure `post-*`, `author-*`, `single-post-*`, `tag-*` show up here
6. Back to templates and open Pages - make sure `page-*` show  up here
7. Back to templates and open General - make sure `$custom` shows up here
8. Back to templates and open Unused - make sure `home`, `index`, and/or `singular` shows up here if conditions are met

## Screenshots 
https://user-images.githubusercontent.com/2256104/108223541-366bac00-713a-11eb-82f8-5481172d68a5.mov

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
